src/storage/src/source/sql_server/replication.rs

```diff
@@ -0,0 +1,337 @@
+                tracing::info!(?needs_snapshot, "starting snapshot");
+                let (snapshot_lsn, snapshot_streams) =
+                    cdc_handle.snapshot(Some(needs_snapshot)).await?;
+                let snapshot_cap = data_cap_set.delayed(&snapshot_lsn);
+
+                let mut snapshot_streams = std::pin::pin!(snapshot_streams);
+                while let Some((capture_instance, data)) = snapshot_streams.next().await {
+                    let sql_server_row = data.map_err(TransientError::from)?;
+
+                    // Decode the SQL Server row into an MZ one.
+                    let Some((partition_idx, decoder)) = capture_instances.get(&capture_instance)
+                    else {
+                        let definite_error = DefiniteError::ProgrammingError(format!(
+                            "capture instance didn't exist: '{capture_instance}'"
+                        ));
+                        let () = return_definite_error(
+                            definite_error,
+                            &output_indexes[..],
+                            data_output,
+                            data_cap_set,
+                            definite_error_handle,
+                            definite_error_cap_set,
+                        )
+                        .await;
+                        return Ok(());
+                    };
+
+                    // Failing to decode data is a permanent failure.
+                    let mut mz_row = Row::default();
+                    let decode_result = decoder.decode(&sql_server_row, &mut mz_row);
+                    tracing::info!(?decode_result, ?mz_row, "snapshotted row");
+                    if let Err(err) = decode_result {
+                        let definite_error = DefiniteError::Decoding(err.to_string());
+                        let () = return_definite_error(
+                            definite_error,
+                            &output_indexes[..],
+                            data_output,
+                            data_cap_set,
+                            definite_error_handle,
+                            definite_error_cap_set,
+                        )
+                        .await;
+                        return Ok(());
+                    }
+
+                    let message = SourceMessage {
+                        key: Row::default(),
+                        value: mz_row,
+                        metadata: Row::default(),
+                    };
+                    data_output
+                        .give_fueled(
+                            &snapshot_cap,
+                            ((*partition_idx, Ok(message)), snapshot_lsn, Diff::ONE),
+                        )
+                        .await;
+                }
+
+                snapshot_lsn
+            };
+
+            // Start replicating from the LSN __after__ we took the snapshot.
+            let replication_start_lsn = snapshot_lsn.increment();
```

Comment:
Is there any documentation to link somewhere in our code that explains exactly what happens during snapshots and whether LSNs are inclusive/exclusive?

Reply:
Added some documentation on the `Lsn` struct and the `mz_sql_server_util::inspect::get_changes` method.